Allow cylc hold to hold future tasks that haven't spawned yet

MetRonnie opened the pull request:
These changes close #3743

You can now make Cylc hold future tasks when they spawn (without having to use the `--after` option).

Note this only works for "exact" task identifiers, e.g. `cylc hold my_flow mytask.1234`; using globs (e.g. `'mytask.*'`) or status qualifiers (e.g. `mytask.1234:failed`) will only hold active tasks, not unspawned ones.

**Requirements check-list**
- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Applied any dependency changes to both `setup.py` and
  `conda-environment.yml`.
- [x] Appropriate tests are included (unit and functional).
- [x] Appropriate change log entry included.
- [x] (master branch) I have opened a documentation PR at cylc/cylc-doc#249.

hjoliver commented on cylc/flow/task_pool.py:
```diff
@@ -705,14 +714,13 @@ def _get_task_by_id(self, id_):
+    def queue_task(self, itask: TaskProxy) -> None:
+        """Queue a task that is ready to run."""
+        itask.state.reset(is_queued=True)
+        # TODO Reset manual trigger flag. One manual trigger queues an
+        # unqueued task, another one triggers a queued task.
```
(FYI this TODO was addressed in #4103 but I evidently neglected to remove the comment ... now done, if you update your branch it will disappear).

hjoliver commented on cylc/flow/task_queues/independent.py:
```diff
@@ -107,15 +106,15 @@ def __init__(self,
         """Release tasks up to the queue limits."""
         released: List[TaskProxy] = []
-        for _, queue in self.queues.items():
+        for queue in self.queues.values():
```
Huh, not sure what I was thinking there! :grimacing:

hjoliver reviewed (changes requested):
You missed one call to `TaskPool.queue_tasks()` that actually queues multiple tasks at once - at the end of config reload - hence the broken functional tests. It's probably fine to queue those tasks one at a time rather than revert your change.

hjoliver commented:
Generally looks great and works well though :tada:  

We probably need a way to show what future tasks are in the hold list. Maybe just add the info to `cylc dump` output.  Or `cylc hold --show` (but that would make `hold` an instant-gratification query command as well as an asynchronous mutation, not sure if that would be a problem...).

MetRonnie commented:
> You missed one call to `TaskPool.queue_tasks()` that actually queues multiple tasks at once - at the end of config reload - hence the broken functional tests. It's probably fine to queue those tasks one at a time rather than revert your change.

Agh, wasn't picked up by mypy because the method calling it wasn't type annotated. Fixed and rebased

MetRonnie commented on cylc/flow/task_pool.py:
```diff
@@ -705,12 +714,11 @@ def _get_task_by_id(self, id_):
-        self.task_queue_mgr.push_tasks(itasks)
+    def queue_task(self, itask: TaskProxy) -> None:
+        """Queue a task that is ready to run."""
+        itask.state.reset(is_queued=True)
+        self.data_store_mgr.delta_task_queued(itask)
+        self.task_queue_mgr.push_task(itask)
```
Might this be a good idea?
```suggestion
        if itask.state.reset(is_queued=True):
            self.data_store_mgr.delta_task_queued(itask)
            self.task_queue_mgr.push_task(itask)
```

hjoliver replied:
It certainly wouldn't hurt.

oliver-sanders commented on cylc/flow/task_pool.py:
```diff
@@ -1002,53 +1008,106 @@ def report_unmet_deps(self):
+                LOG.warning(self.ERR_TMPL_NO_TASKID_MATCH.format(name_str))
+                n_warnings += 1
+                continue
+            if point_str is None:
+                LOG.warning(f"No active instances of task: {item}")
+                n_warnings += 1
+                continue
```
This function is matching future tasks not active ones?

MetRonnie replied:
It's a bit confusing I know. This function relies on you having previously filtered out the matching active tasks using `self.filter_task_proxies()`. So if the point is not specified e.g. `mytask`, equivalent to `mytask.*`, and there is no active instance of `mytask`, then I say that is the correct warning to give.

MetRonnie commented on cylc/flow/workflow_db_mgr.py:
```diff
@@ -496,9 +504,23 @@ def put_task_pool(self, pool: 'TaskPool') -> None:
+                )
                 itask.state.time_updated = None
 
+    def put_tasks_to_hold(
+        self, tasks: Set[Tuple[str, 'PointBase']]
+    ) -> None:
+        """Replace the tasks in the tasks_to_hold table."""
+        # As we are replacing the maps each time this is called, there
+        # shouldn't be much cost to calling this multiple times between
+        # processing of the db queue
+        self.db_deletes_map[self.TABLE_TASKS_TO_HOLD] = [{}]
+        self.db_inserts_map[self.TABLE_TASKS_TO_HOLD] = [
+            {"name": name, "cycle": str(point)}
+            for name, point in tasks
+        ]
```
Hmm, not sure why I did it this way. There may have been a reason why I didn't only insert/delete the differing tasks, but I can't see it

oliver-sanders replied:
I think this might be the approach used for other tables?

For the moment I've got my fingers in my ears going lalala, the DB usage needs some major rationalisation:

* Repeated wiping and rebuilding of tables.
* Repeated creation and destruction of the DB connection.
* Un-normalised tables.

See https://github.com/cylc/cylc-flow/issues/3872, https://github.com/cylc/cylc-flow/issues/3666.

At the moment it's messy but working and we have bigger things to worry about so I think this work will slide back a while to come.

oliver-sanders commented on cylc/flow/task_pool.py:
```diff
@@ -1002,53 +1008,106 @@ def report_unmet_deps(self):
+    def hold_active_task(self, itask: TaskProxy) -> None:
+        if itask.state.reset(is_held=True):
+            self.data_store_mgr.delta_task_held(itask)
+        self.tasks_to_hold.add((itask.tdef.name, itask.point))
```
Active tasks don't need to go into the `tasks_to_hold` set, if they were in the set before I think they should come out?

MetRonnie replied:
They don't need to, but I think it's easier to deal with if everything that is held or is going to be held is in the set. They should only need to come out when released.

oliver-sanders commented on cylc/flow/task_pool.py:
```diff
@@ -1266,33 +1331,45 @@ def spawn_task(
+        task_items: Dict[Tuple[str, 'PointBase'], 'TaskDef'] = {}
         for item in items:
-            point_str, name_str = self._parse_task_item(item)[:2]
+            point_str, name_str, status = self._parse_task_item(item)
```
Status unused?

oliver-sanders reviewed (commented):
Looks great.

MetRonnie commented on cylc/flow/task_pool.py:
```diff
@@ -1004,53 +1012,106 @@ def report_unmet_deps(self):
+        matched_tasks: Set[Tuple[str, 'PointBase']] = set()
+        for item in items:
+            point_str, name_str, status = self._parse_task_item(item)
+            if status or ('*' in item) or ('?' in item) or ('[' in item):
```
I'm not sure what you mean

oliver-sanders replied:
Because globs in the cycle point and name should not be allowed for matching future tasks.

MetRonnie replied:
This checks if a glob character is in any part of the `name[.point][:status]` string

oliver-sanders replied:
Ah, sorry, missread, item includes all three 🤦.

hjoliver reviewed (approved):
Nice, this one is quiet important, for users :+1:

oliver-sanders commented:
Testing is going well, I've hit one issue with the n=1 data store which I had a poke at as I wasn't sure how easy/possible it was to shim, turns out actually quite easy. Here's a PR, feel free to use it or re-work it https://github.com/MetRonnie/cylc-flow/pull/6.

oliver-sanders commented:
I think some tests have been angered by the recent conftest.py change?

oliver-sanders reviewed (approved):
Code looks good, testing went fine, DB housekeeping ok.

Found the following quirks:

* Completed tasks remain in `tasks_to_hold`.

  E.G. If you hold a task then trigger it it will remain in `tasks_to_hold` even after it has passed out of the n=1 window. These stuck tasks can be removed by `cylc release <flow> --all`.

  It's a quirk not a bug, I don't think it's a problem or worth the fuss handling differently.

* Behaviour with reflow.

  I think we will want to add the "flow label" into the `tasks_to_hold` set in order to handle the situation where we have the same task from different flows. E.G. the user might want to hold one flow but the another @hjoliver 

  Easy change to make, happy to do now or later.

* Need to bodge TUI and GUI to handle future families

  > Follow-on not for this PR

  Because globing/matching is not allowed in the n>0 window holding/releasing tasks in the n>0 window from Tui/Gui doesn't work. Easy fix, we should get Tui/Gui to expand the cycle/family to the full list of tasks in the selection.

  Opened:

  * https://github.com/cylc/cylc-flow/issues/4246
  * https://github.com/cylc/cylc-ui/issues/686

oliver-sanders commented:
(will leave @hjoliver with the reflow comment above)

MetRonnie commented:
I have also just noticed:
- If you start a simple workflow (`P1 = foo`) paused, then hold all active tasks, then resume; `foo.1` will apparently ignore the held state and will run

```
INFO - Processing 1 queued command(s)
INFO - Command succeeded: hold(['foo.*'])
INFO - Processing 1 queued command(s)
INFO - RESUMING the workflow now
INFO - Command succeeded: resume()
INFO - Queue released: foo.1
INFO - [foo.1] -submit-num=01, host=hostymchostface
INFO - [foo.1] -triggered off []
INFO - [foo.1] status=preparing (held): (internal)submitted at 2021-06-07  for job(01) flow(E)
INFO - [foo.1] -health check settings: submission timeout=None, polling intervals=PT15M,...
INFO - [foo.1] status=submitted (held): (received)started at 2021-06-07  for job(01) flow(E)
INFO - [foo.1] -health check settings: execution timeout=None, polling intervals=PT15M,...
INFO - [foo.1] status=running (held): (received)succeeded at 2021-06-07  for job(01) flow(E)
```
I think this can be fixed in a separate bugfix PR

MetRonnie commented:
> E.G. If you hold a task then trigger it it will remain in tasks_to_hold even after it has passed out of the n=1 window. These stuck tasks can be removed by `cylc release <flow> --all`.

Hmm, perhaps it should be that triggering a task also releases it?

oliver-sanders commented:
I think we don't do that so that tasks can be triggered in a way that prevents retries from occurring?

There are other ways we could end up with held tasks drifting outside of the `n=1` window (e.g. suicide triggers). I think the solution would be to remove the task from `tasks_to_hold` when the task is removed from the pool.

MetRonnie commented:
Another issue, which I realised from Oliver's comment https://github.com/cylc/cylc-flow/issues/4246#issuecomment-855821083:
- If you have a family `A` that has at cycle point 1:
    - a (n=0)
    - b (n=1)
 
   then doing `cylc hold myflow A.1` will only hold `a.1`, and not set `b.1` to be held. (If `a.1` was n=1 instead, then holding `A.1` would set both `a.1` and `b.1` to be held)

MetRonnie commented:
(Bash test failure is due to Docker container initialisation stalling, don't think it's worth re-running)

hjoliver commented:
> Behaviour with reflow.
> 
> I think we will want to add the "flow label" into the tasks_to_hold set in order to handle the situation where we have the same
> task from different flows. E.G. the user might want to hold one flow but the another @hjoliver

Yeah, I'll post a follow-up issue .... #4277

hjoliver reviewed (approved):
Looks great, no problems found :tada:
